api/types.go

```diff
@@ -707,21 +707,72 @@ func (pT PinType) String() string {
 	}
 }
 
+var pinOptionsMetaPrefix = "meta-"
+
 // PinOptions wraps user-defined options for Pins
 type PinOptions struct {
 	ReplicationFactorMin int               `json:"replication_factor_min" codec:"rn,omitempty"`
 	ReplicationFactorMax int               `json:"replication_factor_max" codec:"rx,omitempty"`
 	Name                 string            `json:"name" codec:"n,omitempty"`
 	ShardSize            uint64            `json:"shard_size" codec:"s,omitempty"`
+	UserAllocations      []string          `json:"user_allocations" codec:"ua,omitempty"`
 	Metadata             map[string]string `json:"metadata" codec:"m,omitempty"`
 }
 
+// Equals returns true of two PinOption objects are equivalent.
+func (po *PinOptions) Equals(po2 *PinOptions) bool {
+	if po.ReplicationFactorMax != po2.ReplicationFactorMax {
+		return false
+	}
+
+	if po.ReplicationFactorMin != po2.ReplicationFactorMin {
+		return false
+	}
+
+	if po.ShardSize != po2.ShardSize {
+		return false
+	}
+
+	lenAllocs1 := len(po.UserAllocations)
+	lenAllocs2 := len(po2.UserAllocations)
+	if lenAllocs1 != lenAllocs2 {
+		return false
+	}
+
+	// avoid side effecs in the original objects
```

Comment:
```suggestion
	// avoid side effects in the original objects
```

May be a naive question, but what kind of side effects?

Reply:
The sorting of the Allocations slice. After calling Equal(), the slice would have changed inadvertently.